### ISSUE TITLE
react/no-unused-prop-types when using prop with styled components

### ISSUE THREAD

**Issue by schoenwaldnils** [yannickcr/eslint-plugin-react#1520]
* **ESLint Version:** 4.10.0
* **Node Version:** 8.4.0
* **npm Version:** 5.5.1

**What parser (default, Babel-ESLint, etc.) are you using?**
babel-eslint ~8.0.1
eslint ~4.10.0
eslint-config-airbnb ~16.0.0
eslint-plugin-import ^2.7.0
eslint-plugin-jsx-a11y ~6.0.2
eslint-plugin-react ^7.1.0

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
  extends: 'airbnb',
  parser: 'babel-eslint',
  plugins: [
    'react'
  ],
  settings: {
    react: {
      'jsx-boolean-value': 'always'
    }
  },
  parserOptions: {
    ecmaFeatures: {
      jsx: true
    }
  },
  rules: {
    'arrow-body-style': 0,
    'jsx-a11y/anchor-is-valid': 0,
    'react/forbid-prop-types': 0,
    'react/jsx-closing-bracket-location': [1, 'after-props'],
    'react/jsx-boolean-value': 0,
    'react/jsx-filename-extension': [
      1,
      {
        'extensions': [
          '.js',
          '.jsx',
        ],
      },
    ],
    'react/no-danger': 0,
    'strict': 0,
    'prefer-template': 0,
    'consistent-return': 0,
    'no-case-declarations': 0,
    'no-console': [
      'warn',
      { allow: [
        'warn',
        'error'
      ]}
    ],
    'no-param-reassign': ['error', { props: false }],
    'no-underscore-dangle': 0,
    'no-prototype-builtins': 0,
  },
  env: {
    browser: true
  }
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
Picture.propTypes = {
  width: PropTypes.number, // 'width' PropType is defined but prop is never used (react/no-unused-prop-types)
}
export default styled(Component)`
  width: ${props => `${props.width}px`};
`;
```

**What did you expect to happen?**
eslint should not throw an error if the prop is used with __styled components__

**What actually happened? Please include the actual, raw output from ESLint.**
```
  68:10  error  'width' PropType is defined but prop is never used  react/no-unused-prop-types

✖ 1 problem (1 error, 0 warnings)
```

Moved here from https://github.com/eslint/eslint/issues/9593

**Comment by ljharb** [comment]
The pattern you're using - a tagged template string - is very strange, magic, hard to optimize, and impossible to lint reliably.

Having a linter plugin that can parse the template literal to do static analysis on it is something that I'd say belongs in a "styled components"-specific linter plugin, and definitely not in a broadly React-focused plugin.

**Comment by schoenwaldnils** [comment]
This is the default implementation from styled-components, like you can see here: https://www.styled-components.com/

Shouldn't it be possible to parse the tagged template literal to see if a prop is used or not?

**Comment by ljharb** [comment]
I'm sure it's possible, yes - but that's a lot of complexity that a plugin for React shouldn't have to deal with - a plugin specifically for styled-components would have to deal with it.

**Comment by kpavankotesh** [comment]
Same problem while using [CssinJS](http://cssinjs.org/react-jss ). Eslint should check for props used in styles as well

```
const styles = {
  button: {
    background: props => props.color
  }
}
```

**Comment by ljharb** [comment]
@kpavankotesh that's just a random function; an eslint plugin has no reliable way to know for sure that you invoked it with `this.props`;

**Comment by r3sMetz** [comment]
idk if its still relevant but i solved it giving the styled-component its own propTypes and defaultProps

```
// Component
const Hero = ({ className, headline, subline }) => (
  <header className={className}>
    <Container>
      {headline && <HeadingOne>{headline}</HeadingOne>}
      {subline && <HeadingThree>{subline}</HeadingThree>}
    </Container>
  </header>
)

// Component PropTypes
Hero.propTypes = {
  className: PropTypes.string,
  headline: PropTypes.string,
  subline: PropTypes.string,
}

// Styling
const StyledHero = styled(Hero)`
  text-align: center;
  background-size: cover;
  background-image: url(${props => props.bgImagePath});
  background-position: center;
  color: ${colors.defaults.white};
  ${Container} {
    padding-top: ${spacing[5]};
    padding-bottom: ${spacing[5]};
  }
`

// Styling PropTypes + Default Props
StyledHero.propTypes = {
  bgImagePath: PropTypes.string,
}
StyledHero.defaultProps = {
  bgImagePath: '/header-bg.jpg',
}

export default StyledHero
```
